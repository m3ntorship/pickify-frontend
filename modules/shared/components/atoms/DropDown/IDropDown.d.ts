declare namespace IDropDown {
  export interface IProps extends IHandlers {
    options: IOptions[];
    variant: 'image' | 'post';
    size: 'md' | 'sm';
    isAuthenticated?: boolean;
  }

  export interface IOptions {
    body: string;
    id: string;
  }

  export interface IHandlers {
    onOptionMenuClick: (id: string) => void;
  }
}

export { IDropDown };
