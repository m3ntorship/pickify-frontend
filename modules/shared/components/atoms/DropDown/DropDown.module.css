.menu-icon-container {
  @apply cursor-pointer bg-white;
}

.menu-icon {
  @apply fill-grey w-6 h-6;
}

.menu {
  @apply bg-white shadow-soft rounded-md p-2 flex flex-col absolute -top-2 right-8 md:left-12 w-5xl;
}

.menu-item {
  @apply bg-white hover:bg-grey-bg py-1 px-2 rounded-sm cursor-pointer hover:text-dark text-dark-grey text-sm;
}
