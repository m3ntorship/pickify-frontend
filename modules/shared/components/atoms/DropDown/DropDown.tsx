import React, { useState } from 'react';

import type { FC, ReactElement } from 'react';
import type { IDropDown } from './IDropDown';
import VerticalThreeDotsIcon from '../../icons/verticalThreeDots.svg';
import styles from './DropDown.module.css';

const DropDown: FC<IDropDown.IProps> = ({
  options,
  onOptionMenuClick,
}): ReactElement => {
  const [showMenu, setShowMenu] = useState<boolean>(false);

  const showMenuHandler = (): void => {
    setShowMenu(!showMenu);
  };

  return (
    <div className="flex">
      <VerticalThreeDotsIcon
        onClick={showMenuHandler}
        className={styles['menu-icon']}
        data-testid="menu-icon"
      />
      {showMenu && (
        <div role="menu" className={styles.menu} data-testid="menu">
          {options.map((option) => (
            <button
              className={styles['menu-item']}
              role="menuitem"
              type="button"
              name="option"
              id={option.id}
              key={option.id}
              onClick={onOptionMenuClick}
              data-testid="menu-button"
            >
              {option.body}
            </button>
          ))}
        </div>
      )}
    </div>
  );
};

export default DropDown;
