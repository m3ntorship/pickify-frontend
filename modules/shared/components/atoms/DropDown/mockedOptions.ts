export const options = [
  { id: '1', body: 'Delete Post' },
  { id: '1', body: 'Report Post' },
  { id: '1', body: 'Save Post' },
];
