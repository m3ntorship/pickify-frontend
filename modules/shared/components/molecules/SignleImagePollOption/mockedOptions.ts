export const covoredOptions = [
  { body: 'yes', id: 'option_1', media: [] },
  { body: 'no', id: 'option_2', media: [] },
];

export const unCovoredOptions = [
  { body: 'yes', id: 'option_1', media: [], vote_count: 3 },
  { body: 'no', id: 'option_2', media: [], vote_count: 1 },
];
