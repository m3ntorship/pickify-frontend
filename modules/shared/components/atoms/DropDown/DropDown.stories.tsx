import type { ReactElement } from 'react';
import type { Story } from '@storybook/react';
import DropDown from './DropDown';
import type { IDropDown } from './IDropDown';
import { options } from './mockedOptions';

export default {
  title: 'Atoms/DropDown',
  component: DropDown,
};

const Template: Story<IDropDown.IProps> = (args): ReactElement => (
  <div className="flex justify-end">
    <DropDown {...args} />
  </div>
);
export const postViewHeader = Template.bind({});
postViewHeader.args = {
  options,
};
