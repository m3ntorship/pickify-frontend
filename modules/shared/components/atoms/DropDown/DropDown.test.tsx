import React from 'react';
import { render, screen } from '@testing-library/react';
import type { TargetElement } from '@testing-library/user-event';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import DropDown from './DropDown';
import { options } from './mockedOptions';

describe('DropDown', () => {
  it('should render the only menu icon', () => {
    const mockedFn = jest.fn();
    const tree = renderer
      .create(<DropDown onOptionMenuClick={mockedFn} options={options} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render three menu items when passing three options and clicking on menu icon', () => {
    const threeMenuItems = 3;

    const mockedFn = jest.fn();
    render(<DropDown onOptionMenuClick={mockedFn} options={options} />);

    const menuIcon: TargetElement = screen.getByTestId('menu-icon');

    userEvent.click(menuIcon);

    const menuButtons: TargetElement[] = screen.getAllByTestId('menu-button');
    const menu: TargetElement = screen.getByTestId('menu');

    expect(menu).toBeInTheDocument();
    expect(menuButtons).toHaveLength(threeMenuItems);
  });

  it('should call onOptionMenuClick when clicking on the menu button', () => {
    const firstButton = 0;
    const calledOnce = 1;

    const mockedFn = jest.fn();
    render(<DropDown onOptionMenuClick={mockedFn} options={options} />);

    const menuIcon: TargetElement = screen.getByTestId('menu-icon');

    userEvent.click(menuIcon);

    const menuButtons: TargetElement[] = screen.getAllByTestId('menu-button');

    userEvent.click(menuButtons[firstButton]);

    expect(mockedFn).toHaveBeenCalledTimes(calledOnce);
  });
});
