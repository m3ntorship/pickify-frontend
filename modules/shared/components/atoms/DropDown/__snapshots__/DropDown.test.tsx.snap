// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown should render the only menu icon 1`] = `
<div
  className="flex"
>
  <svg
    className="menu-icon"
    data-file-name="SvgVerticalThreeDots"
    data-testid="menu-icon"
    onClick={[Function]}
  />
</div>
`;
